core/psh/psh.c

```diff
@@ -972,7 +972,181 @@ static void psh_signalstop(int sig)
+
+	/* Get login from user */
+	memset(indata,'\0',16);
+	for (;;) {
+		if (insz < 0) {
+			fprintf(stdout,"Login: ");
+			fflush(0);
+			insz++;
+		}
+
+		read(STDIN_FILENO, &c, 1);
+		if (c == '\n') {
+			fprintf(stdout,"\n");
+			if (insz <= maxinsz && insz > 0) {
+				memset(indatasave,'\0',17);
+				memcpy(indatasave,indata,sizeof(char)*16);
+				memset(indata,'\0',17);
+				break;
+			}
+			else {
+				memset(indata,'\0',16);
+				insz = -1;
+				continue;
+			}
+		}
+		else if (c > 31 && c < 127) {
+			fprintf(stdout,"%c",c);
+			fflush(stdout);
+			if (insz < maxinsz)
+				indata[insz] = c;
+			insz++;
+			continue;
+		}
+		else if (c == 8 || c==127) {
+			if (insz > 0) {
+				insz--;
+				fprintf(stdout,"\b \b");
+				fflush(stdout);
+				if (insz < maxinsz)
+					indata[insz] = '\0';
+			}
+		}
+	}
+
+	/* Get password from user */
+	memset(shdw,'\0',31);
+	memset(indata,'\0',16);
+	insz = -1;
+	for (;;) {
+		if (insz < 0) {
+			printf("Password: ");
+			fflush(0);
+			insz++;
+		}
+
+		read(STDIN_FILENO, &c, 1);
+		if (c == '\n') {
+			memset(shdw,'\0',12);
+			if (insz <= maxinsz) {
+				memcpy(shdw,crypt(indata,"ABCD"),sizeof(char)*31); 
+				memset(indata,'\0',16);
+				insz = 0;
+				break;
+			}
+			else {
+				memset(indata,'\0',16);
+				insz = -1;
+				break;
+			}
+		}
+		else if (c > 31 && c < 127) {
+			if (insz < maxinsz)
+				indata[insz] = c;
+			insz++;
+		}
+		else if (c == 8) {
+			if (insz > 0) {
+				insz--;
+				if (insz < maxinsz) {
+					indata[insz] = '\0';
+				}
+			}
+		}
+	}
+	fprintf(stdout,"\n");
+	mode.c_lflag |= (ECHO | ICANON);
+	tcsetattr(0, TCSANOW, &mode);
+
+	/* Try to access /etc/passwd file */
+	FILE *shadows = NULL;
+	if (access("/etc/passwd",R_OK) == 0) {
+		shadows = fopen("/etc/passwd","r");
+		while ((c = fgetc(shadows)) != EOF) {
+			if (c=='\n')
+				shdwlines++;
+		}
+		rewind(shadows);
+		if (shdwlines == 0) {
+			fclose(shadows);
+			shadows = NULL;
+		} 
+	}
+
+	/* Compare login and shadow against /etc/passwd */
+	code = -EACCES;
+	if (shadows != NULL) {
+		for (int ii = 0; ii <= shdwlines; ii++) {
+			insz = 0;
+			memset(indata,'\0',16);
+			while ((c = fgetc(shadows)) != ':' && c > 0) {
+				if (insz < maxinsz)
+					indata[insz] = c;
+				insz++;
+			}
+			/* if login matches then check password */
+			code = -EACCES;
+			if (strcmp(indata,indatasave) ==0 && (insz <= maxinsz)) {
+				c = fgetc(shadows);
+				int gg=0;
+				while (c > 0 && c != ':') {
+					if (c != shdw[gg++]) {
+						code = -EACCES;
+						break;
+					}
+					c = fgetc(shadows);
+				}
+				/* success exit only if shadow matches or shadow is empty */
+				if (c == ':' && (shdw[gg] == '\0' || gg == 0)) {
+					code = 0;
+					break;
+				}
+			}
+			memset(indata,'\0',16);
+
+			/* File exit decision */
+			if (code != 0) {
+				while (c != '\n' && c != EOF)
+					c = fgetc(shadows);
+				code = -EACCES;
+				continue;
+			}
+			else {
+				break;
+			}
+		}
+		fclose(shadows);
+	}
+
+	#ifdef DEFUSR_PASSWD
```

Comment:
This is deliberate (and consulted) to give an option of passing password value to system implementations which do not have root filesystem, thus no /etc/passwd file. This gives an opportunity to add password to system without storing it hardcoded.

Reply:
- we  for sure need `PSH_` prefix not to pollute compile environment
- it will still be hardcoded in the result binary, just can be changed at compile time
- not sure if '#ifdef' should be indented

Reply:
It's nice that we will store the password hash instead of the password in plaintext now.